src/clapsolver/dynamic_lsap.py

```diff
@@ -0,0 +1,261 @@
+    sub_cost_matrix = cost_matrix[:, col4row]
+    sub_v = v[col4row]
+
+    # Remove the assignment associated with the removed row. Note that in the
+    # submatrix, the removed row and the freed up column have the same index.
+    # TODO: check whether the following expressions are correct
+    sub_col4row = np.arange(n_rows)
+    sub_col4row[row_removed] = -1
+    sub_row4col = np.arange(n_rows)
+    sub_row4col[row_removed] = -1
+
+    # Find the shortest augmenting path for the sub square lsap and augment.
+    augment(sub_cost_matrix, row_removed, sub_row4col, sub_col4row, u, sub_v)
+
+    # Update the original assignment
+    # Note: update every variable that depends on col4row for indexing first.
+    row4col[col4row] = sub_row4col
+    v[col4row] = sub_v
+    col4row = col4row[sub_col4row]
+
+    return row4col, col4row, u, v
+
+
+def linear_sum_assignment(cost_matrix, maximize=False):
+    """Solve the linear sum assignment based on the cost matrix. The return
+       value is the same as in scipy.
+
+    Parameters
+    ----------
+    cost_matrix : 2darray
+         A matrix of costs.
+
+    Returns
+    -------
+    row_ind, col_ind : array
+        An array of row indices and one of corresponding column indices giving
+        the optimal assignment. The cost of the assignment can be computed
+        as ``cost_matrix[row_ind, col_ind].sum()``. The row indices will be
+        sorted; in the case of a square cost matrix they will be equal to
+        ``numpy.arange(cost_matrix.shape[0])``.
+    """
+    cost_matrix = np.array(cost_matrix)
+
+    # The following are taken from scipy implementation.
+    if len(cost_matrix.shape) != 2:
+        raise ValueError(
+            "expected a matrix (2-d array), got a %r array" % (cost_matrix.shape,)
+        )
+
+    # fmt: off
+    # TODO: the following is not passing flake8
+    if not (np.issubdtype(cost_matrix.dtype, np.number) or
+            cost_matrix.dtype == np.dtype(np.bool)):
+        raise ValueError(
+            "expected a matrix containing numerical entries, got %s"
+            % (cost_matrix.dtype,)
+        )
+    # fmt: on
+
+    if maximize:
+        cost_matrix = -cost_matrix
+
+    if np.any(np.isneginf(cost_matrix) | np.isnan(cost_matrix)):
+        raise ValueError("matrix contains invalid numeric entries")
+
+    # cost_matrix = cost_matrix.astype(np.double)
+    a = np.arange(np.min(cost_matrix.shape))
+
+    # If the cost_matrix has more rows than columns
+    if cost_matrix.shape[1] < cost_matrix.shape[0]:
+        # Here, col4row holds the rows in cost_matrix that are in the assignment
+        _, col4row, _, _ = solve_lsap(cost_matrix.T)
+
+        # Sort the row indexes in the assignment
+        idx_sorted = np.argsort(col4row)
+
+        return (col4row[idx_sorted], a[idx_sorted])
+    # If the cost_matrix has more columns than rows
+    else:
+        _, col4row, _, _ = solve_lsap(cost_matrix)
+        return (a, col4row)
```

Comment:
This indicates that the `solve_lsap` function requires the `cost_matrix` to have certain shape. Make sure the allowed shape is clearly mentioned in the docstring of `solve_lsap`.